llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp

```diff
@@ -2916,8 +2916,11 @@ void CombinerHelper::replaceInstWithFConstant(MachineInstr &MI, double C) {
 
 void CombinerHelper::replaceInstWithConstant(MachineInstr &MI, int64_t C) {
   assert(MI.getNumDefs() == 1 && "Expected only one def?");
-  Builder.buildConstant(MI.getOperand(0), C);
-  MI.eraseFromParent();
+  LLT DstTy = MRI.getType(MI.getOperand(0).getReg());
+  if (isLegalOrBeforeLegalizer({TargetOpcode::G_CONSTANT, {DstTy}})) {
+    Builder.buildConstant(MI.getOperand(0), C);
+    MI.eraseFromParent();
+  }
```

Comment:
Please take a step back to address the root causes of the issues.

We can't be just not doing what the function says it would do. This change needs its own rationale or PR if it is indeed the right change to do.

Reply:
I am interested in root cause analysis. We are combining G_ZEXT of undef into 0 in this PR and it relies on this function. We never build something before checking legality. The change itself is not negotiable. You can ask to move into a separate PR of course.